README.md
# Slide Digital Clock Plugin

## Usage

To use plugin, just import package `import 'package:slide_digital_clock/slide_digital_clock.dart';`

## Defaults

    is24HourTimeFormat : `true`
    showSecondsDigit : `true`  
    areaAligment : `AlignmentDirectional.bottomCenter`   
    digitAnimationStyle : `Curves.decelerate`
    hourMinuteDigitTextStyle : `Theme.of(context).textTheme.bodyText1`
    secondDigitTextStyle : `Theme.of(context).textTheme.caption!.copyWith(fontSize: 10))`
    amPmDigitTextStyle : `Theme.of(context).textTheme.caption!.copyWith(fontSize: 10, fontWeight: FontWeight.bold)`

## Demo

![Demo: ](demo.gif)

## Pics

![Demo: ](example.png)

## Example

```
import 'package:flutter/material.dart';
import 'package:slide_digital_clock/slide_digital_clock.dart';

void main() {
  runApp(const ExampleApp());
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: "Slider Digital Clock Demo App",
      home: Scaffold(
        appBar: AppBar(
            title: const Text("Slider Digital Clock Demo App"),
            actions: [
              DigitalClock(
                hourMinuteDigitTextStyle: Theme.of(context)
                    .textTheme
                    .headline4!
                    .copyWith(color: Colors.white),
                secondDigitTextStyle: Theme.of(context)
                    .textTheme
                    .caption!
                    .copyWith(color: Colors.white),
                colon: Text(
                  ":",
                  style: Theme.of(context)
                      .textTheme
                      .subtitle1!
                      .copyWith(color: Colors.white),
                ),
              )
            ]),
        body: Column(
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Custom Colon"),
                const SizedBox(width: 50),
                DigitalClock(
                  hourMinuteDigitTextStyle: const TextStyle(fontSize: 18),
                  colon: const Icon(Icons.ac_unit, size: 12),
                  colonDecoration: BoxDecoration(
                      border: Border.all(), shape: BoxShape.circle),
                )
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Custom Styles"),
                const SizedBox(width: 50),
                DigitalClock(
                  hourDigitDecoration: BoxDecoration(
                      color: Colors.yellow,
                      border: Border.all(color: Colors.blue, width: 2)),
                  minuteDigitDecoration: BoxDecoration(
                      color: Colors.yellow,
                      border: Border.all(color: Colors.red, width: 2)),
                  secondDigitDecoration: BoxDecoration(
                      color: Colors.blueGrey,
                      border: Border.all(color: Colors.blue),
                      shape: BoxShape.circle),
                  secondDigitTextStyle: Theme.of(context)
                      .textTheme
                      .caption!
                      .copyWith(color: Colors.white),
                ),
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Show AM-PM, Disable Seconds"),
                const SizedBox(width: 50),
                DigitalClock(
                  is24HourTimeFormat: false,
                  showSecondsDigit: false,
                ),
              ],
            )
          ],
        ),
      ),
    );
  }
}


```


CHANGELOG.md
## [1.0.3] : 28.01.2023

* New animation for spinner text
* Customizable decoration for all digits
* Add animation for colon

## [1.0.2] : 11.08.2020

* Fix Null-Safety bugs

## [1.0.1] : 11.08.2020

* Fix Null-Safety

## [1.0.0] : 06.01.2020

* TODO : bottom-up animation to be added
* Added example project

## [0.0.3] : 06.01.2020

* TODO : bottom-up animation to be added

### What's Fixed

* Fixed some text files

## [0.0.2] : 06.01.2020

* TODO : bottom animation to be added
* Fixed issue requiring am-pm
* Automatic adjustment of field dimensions according to Font size of Hour-Minute Text

## [0.0.1] : 05.01.2020

* This release contains the 'DigitalClock' class was defined 'slider'.


test/slide_digital_clock_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:slide_digital_clock/src/digital_clock.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(DigitalClock());
  });
}


lib/src/digital_clock.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'helpers/colon.dart';

import 'helpers/clock_model.dart';
import 'helpers/spinner_text.dart';

class DigitalClock extends StatefulWidget {
  DigitalClock({
    this.is24HourTimeFormat,
    this.showSecondsDigit,
    this.colon,
    this.colonDecoration,
    this.areaWidth,
    this.areaHeight,
    this.areaDecoration,
    this.areaAligment,
    this.hourDigitDecoration,
    this.minuteDigitDecoration,
    this.secondDigitDecoration,
    this.digitAnimationStyle,
    this.hourMinuteDigitTextStyle,
    this.secondDigitTextStyle,
    this.amPmDigitTextStyle,
  });

  /// am or pm
  final bool? is24HourTimeFormat;

  /// if you want use seconds this variable should be true
  final bool? showSecondsDigit;

  /// use ":"  or create your widget
  final Widget? colon;
  // colon area decoraiton
  final BoxDecoration? colonDecoration;

  /// clock area width
  final double? areaWidth;

  ///clock area height
  final double? areaHeight;

  /// clock area decoration
  final BoxDecoration? areaDecoration;

  final AlignmentDirectional? areaAligment;

  /// hour decoration
  final BoxDecoration? hourDigitDecoration;

  /// minute decoration
  final BoxDecoration? minuteDigitDecoration;

  /// seconds decoration
  final BoxDecoration? secondDigitDecoration;

  /// animation style
  final Curve? digitAnimationStyle;

  /// hour text style
  final TextStyle? hourMinuteDigitTextStyle;

  /// seconds text style,
  final TextStyle? secondDigitTextStyle;

  /// am-pm text style
  final TextStyle? amPmDigitTextStyle;

  @override
  _DigitalClockState createState() => _DigitalClockState();
}

class _DigitalClockState extends State<DigitalClock> {
  late DateTime _dateTime;
  late ClockModel _clockModel;
  late Timer _timer;

  @override
  void initState() {
    super.initState();
    _dateTime = DateTime.now();
    _clockModel = ClockModel();
    _clockModel.is24HourFormat = widget.is24HourTimeFormat ?? true;

    _dateTime = DateTime.now();
    _clockModel.hour = _dateTime.hour;
    _clockModel.minute = _dateTime.minute;
    _clockModel.second = _dateTime.second;

    _timer = Timer.periodic(Duration(seconds: 1), (timer) {
      _dateTime = DateTime.now();
      _clockModel.hour = _dateTime.hour;
      _clockModel.minute = _dateTime.minute;
      _clockModel.second = _dateTime.second;

      setState(() {});
    });
  }

  @override
  void dispose() {
    _timer.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(4.0),
      child: SizedBox(
        width: widget.areaWidth ?? null,
        height: widget.areaHeight ?? null,
        child: Container(
          alignment: widget.areaAligment ?? AlignmentDirectional.bottomCenter,
          decoration: widget.areaDecoration,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              _amPm,
              _hour(),
              Container(
                alignment: AlignmentDirectional.center,
                margin: const EdgeInsets.all(1.0),
                padding: const EdgeInsets.all(2.0),
                decoration: widget.colonDecoration,
                child: ColonWidget(
                  colon: widget.colon,
                ),
              ),
              _minute,
              _second,
            ],
          ),
        ),
      ),
    );
  }

  Widget _hour() => Container(
        padding: EdgeInsets.all(2),
        alignment: AlignmentDirectional.center,
        decoration: widget.hourDigitDecoration,
        child: SpinnerText(
          text: _clockModel.is24HourTimeFormat
              ? hTOhh_24hTrue(_clockModel.hour)
              : hTOhh_24hFalse(_clockModel.hour)[0],
          animationStyle: widget.digitAnimationStyle,
          textStyle: widget.hourMinuteDigitTextStyle ??
              Theme.of(context).textTheme.bodyText1,
        ),
      );

  Widget get _minute => Container(
        padding: EdgeInsets.all(2),
        alignment: AlignmentDirectional.center,
        decoration: widget.minuteDigitDecoration,
        child: SpinnerText(
          text: mTOmm(_clockModel.minute),
          animationStyle: widget.digitAnimationStyle,
          textStyle: widget.hourMinuteDigitTextStyle ??
              Theme.of(context).textTheme.bodyText1,
        ),
      );

  Widget get _second => widget.showSecondsDigit != false
      ? Container(
          margin: EdgeInsets.all(1),
          padding: EdgeInsets.all(2),
          alignment: AlignmentDirectional.center,
          decoration: widget.secondDigitDecoration,
          child: SpinnerText(
              text: sTOss(_clockModel.second),
              animationStyle: widget.digitAnimationStyle,
              textStyle: widget.secondDigitTextStyle ??
                  Theme.of(context).textTheme.caption!.copyWith(fontSize: 10)),
        )
      : SizedBox();

  Widget get _amPm => _clockModel.is24HourTimeFormat
      ? SizedBox()
      : Container(
          padding: EdgeInsets.all(2),
          alignment: AlignmentDirectional.center,
          child: Text(
            " " + hTOhh_24hFalse(_clockModel.hour)[1],
            style: widget.amPmDigitTextStyle ??
                Theme.of(context)
                    .textTheme
                    .caption!
                    .copyWith(fontSize: 10, fontWeight: FontWeight.bold),
          ),
        );
}


lib/src/helpers/colon.dart
import 'dart:async';

import 'package:flutter/material.dart';

class ColonWidget extends StatefulWidget {
  final Widget? colon;

  const ColonWidget({Key? key, this.colon}) : super(key: key);

  @override
  State<ColonWidget> createState() => _ColonWidgetState();
}

class _ColonWidgetState extends State<ColonWidget> {
  late Timer timer;
  bool visible = true;
  @override
  void initState() {
    super.initState();
    timer = Timer.periodic(Duration(milliseconds: 500), (timer) {
      setState(() {
        visible = !visible;
      });
    });
  }

  @override
  void dispose() {
    super.dispose();
    timer.cancel();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedOpacity(
      opacity: visible ? 1.0 : 0.0,
      duration: const Duration(milliseconds: 500),
      child: widget.colon ??
          Text(
            ":",
            style: Theme.of(context).textTheme.caption,
          ),
    );
  }
}


lib/helpers/clock_model.dart
class ClockModel {

  late int hour;
  late int minute;
  late int second;
  late bool is24HourFormat;

  get is24HourTimeFormat => this.is24HourFormat;
}

hTOhh_24hTrue(int hour) {
  late String sHour;
  if (hour < 10) {
    sHour = "0$hour";
  } else {
    sHour = "$hour";
  }
  return sHour;
}

hTOhh_24hFalse(int hour) {
  late String sHour;
  late String h12State;
  var times = [];
  if (hour < 10) {
    sHour = "0$hour";
    h12State = "AM";
  } else if (hour > 9 && hour < 12) {
    sHour = "$hour";
    h12State = "AM";
  } else if(hour == 12) {
     sHour = "12";
    h12State = "PM";
  } else if (hour > 12 && hour < 22) {
    sHour = "0${hour % 12}";
    h12State = "PM";
  } else if (hour > 21) {
    sHour = "${hour % 12}";
    h12State = "PM";
  }
  times.add(sHour);
  times.add(h12State);
  return times;
}

mTOmm(int minute) {
  late String sMinute;
  if (minute < 10) {
    sMinute = "0$minute";
  } else {
    sMinute = "$minute";
  }
  return sMinute;
}

sTOss(int second) {
  late String sSecond;
  if (second < 10) {
    sSecond = "0$second";
  } else {
    sSecond = "$second";
  }
  return sSecond;
}


lib/src/helpers/spinner_text.dart
import 'package:flutter/material.dart';

class SpinnerText extends StatefulWidget {
  SpinnerText({required this.text, this.textStyle, this.animationStyle});

  final String text;
  final TextStyle? textStyle;
  final Curve? animationStyle;

  _SpinnerTextState createState() => _SpinnerTextState();
}

class _SpinnerTextState extends State<SpinnerText>
    with SingleTickerProviderStateMixin {
  String topText = "";
  String bottomText = "";

  late AnimationController _spinTextAnimationController;
  late Animation<double> _spinAnimation;

  @override
  void initState() {
    super.initState();
    bottomText = widget.text;
    _spinTextAnimationController = new AnimationController(
        duration: const Duration(milliseconds: 500), vsync: this)
      ..addListener(() => setState(() {}))
      ..addStatusListener((AnimationStatus status) {
        if (status == AnimationStatus.completed) {
          setState(() {
            bottomText = topText;
            topText = "";
            _spinTextAnimationController.value = 0.0;
          });
        }
      });

    _spinAnimation = CurvedAnimation(
        parent: _spinTextAnimationController,
        curve: widget.animationStyle ?? Curves.ease);
  }

  @override
  void dispose() {
    _spinTextAnimationController.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(SpinnerText oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (widget.text != oldWidget.text) {
      //Need to spin new value
      topText = widget.text;
      _spinTextAnimationController.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    return ClipRect(
      clipper: RectClipper(),
      child: Stack(
        children: <Widget>[
          FractionalTranslation(
            translation: Offset(0.0, 1 - _spinAnimation.value),
            child: Text(
              topText,
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
              style: widget.textStyle,
            ),
          ),
          FractionalTranslation(
            translation: Offset(0.0, _spinAnimation.value),
            child: Text(bottomText,
                maxLines: 1,
                overflow: TextOverflow.ellipsis,
                style: widget.textStyle),
          ),
        ],
      ),
    );
  }
}

class RectClipper extends CustomClipper<Rect> {
  @override
  Rect getClip(Size size) {
    return Rect.fromLTWH(0.0, 0.0, size.width, size.height + 1);
  }

  @override
  bool shouldReclip(CustomClipper<Rect> oldClipper) {
    return true;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slide_digital_clock/slide_digital_clock.dart';

void main() {
  runApp(const ExampleApp());
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: "Slider Digital Clock Demo App",
      home: Scaffold(
        appBar: AppBar(
            title: const Text("Slider Digital Clock Demo App"),
            actions: [
              DigitalClock(
                hourMinuteDigitTextStyle: Theme.of(context)
                    .textTheme
                    .headline4!
                    .copyWith(color: Colors.white),
                secondDigitTextStyle: Theme.of(context)
                    .textTheme
                    .caption!
                    .copyWith(color: Colors.white),
                colon: Text(
                  ":",
                  style: Theme.of(context)
                      .textTheme
                      .subtitle1!
                      .copyWith(color: Colors.white),
                ),
              )
            ]),
        body: Column(
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Custom Colon"),
                const SizedBox(width: 50),
                DigitalClock(
                  hourMinuteDigitTextStyle: const TextStyle(fontSize: 18),
                  colon: const Icon(Icons.ac_unit, size: 12),
                  colonDecoration: BoxDecoration(
                      border: Border.all(), shape: BoxShape.circle),
                )
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Custom Styles"),
                const SizedBox(width: 50),
                DigitalClock(
                  hourDigitDecoration: BoxDecoration(
                      color: Colors.yellow,
                      border: Border.all(color: Colors.blue, width: 2)),
                  minuteDigitDecoration: BoxDecoration(
                      color: Colors.yellow,
                      border: Border.all(color: Colors.red, width: 2)),
                  secondDigitDecoration: BoxDecoration(
                      color: Colors.blueGrey,
                      border: Border.all(color: Colors.blue),
                      shape: BoxShape.circle),
                  secondDigitTextStyle: Theme.of(context)
                      .textTheme
                      .caption!
                      .copyWith(color: Colors.white),
                ),
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Custom Styles"),
                const SizedBox(width: 50),
                DigitalClock(
                  hourDigitDecoration: const BoxDecoration(
                      color: Colors.yellow, shape: BoxShape.circle),
                  minuteDigitDecoration: const BoxDecoration(
                      color: Colors.yellow, shape: BoxShape.circle),
                  secondDigitDecoration: BoxDecoration(
                      color: Colors.blueGrey,
                      border: Border.all(color: Colors.blue),
                      shape: BoxShape.circle),
                  secondDigitTextStyle: Theme.of(context)
                      .textTheme
                      .caption!
                      .copyWith(color: Colors.white),
                  colonDecoration: BoxDecoration(
                      border: Border.all(color: Colors.blue, width: 1),
                      shape: BoxShape.circle),
                ),
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                const Text("Show AM-PM, Disable Seconds"),
                const SizedBox(width: 50),
                DigitalClock(
                  is24HourTimeFormat: false,
                  showSecondsDigit: false,
                ),
              ],
            )
          ],
        ),
      ),
    );
  }
}
